### PR TITLE
Merge diagnosticsProducing and nonDiagnosticsProducing checkers into a single checker supporting lazy diagnostics

### LINKED ISSUE #27997: Expose getDiagnosticsProducingTypeChecker to avoid duplicate type instantiations
## Search Terms

diagnosticsProducingTypeChecker
TypeChecker
getTypeChecker

## Suggestion

Expose [`getDiagnosticsProducingTypeChecker`](https://github.com/Microsoft/TypeScript/blob/91bb32ade40cd2fdf35b32fb13fff5b71c123fe2/src/compiler/types.ts#L2878).

## Use Cases

A custom compiler adding additional checks must currently call:
```typescript
// program: typescript.Program
const originalDiagnostics = program.get*Diagnostics(sourceFile);
const typeChecker: typescript.TypeChecker = program.getTypeChecker();
/* Use `typeChecker` to check the source file for additional diagnostics. */
```

Internally, [`get*Diagnostics` methods use `getDiagnosticsProducingTypeChecker`](https://github.com/Microsoft/TypeScript/blob/e4718564e5b2c1f2e7bbef27fc480fd8172dfdc0/src/compiler/program.ts#L1500), so it is already "warmed" - types are already instantiated. The current implementation requires that a compiler pay the cost to instantiate types twice: once with a `diagnosticsProducingTypeChecker`, and once with a `noDiagnosticsTypeChecker`.

This would eliminate the redundant warm-up cost for instantiating types in the `TypeChecker`. This would be especially effective in code bases with deeply nested type definitions which may take noticeably longer to instantiate.

I do not fully understand the difference between the two `TypeCheckers`, so I cannot honestly comment on the drawbacks to exposing/using a `diagnosticsProducingTypeChecker` for custom checks (*can someone please explain this, or point me towards documentation?*). If it is unfeasible, perhaps there is a way to warm both checkers on the same run, to reduce the repeated type instantiations.

## Examples

See Use Cases.

## Checklist

My suggestion meets these guidelines:
* [X] This wouldn't be a breaking change in existing TypeScript / JavaScript code
* [X] This wouldn't change the runtime behavior of existing JavaScript code
* [X] This could be implemented without emitting different JS based on the types of the expressions
* [X] This isn't a runtime feature (e.g. new expression-level syntax)

### PR DESCRIPTION
**Pull request by weswigham**
A more complete alternative to #28584 based on the discussions there.

Rather than optimistically using an existing checker, we choose to only ever produce a single checker. That checker is responsible for both collecting diagnostics, and responding to LS queries. It does this (hopefully efficiently) by deferring calculating diagnostics for any nodes it visited during LS operations until they are later requested. This allows results from a "full typecheck" (or `getErr` request) to then inform a following completions or quickinfo request, without needing to recalculate anything that's already been calculated. Plus, we no longer need to hold onto 2 checkers in services, which should commensurately lower services memory usage.

To be as minimally breaky as possible, in situations where we know we're about to fetch diagnostics, we produce the diagnostics as they're registered, in order to preserve existing union ordering (as union order is based upon order of discovery, which may change if diagnostic production is fully unconditionally deferred until after checking is complete). This means command-line compiler runs before and after this PR should be identical (as nothing has changed when it executes). What _is_ deferred is any diagnostics we produce while doing LS operations in the checker before we request diagnostics. These lazy diagnostic requests are backlogged until a diagnostic (or suggestion) request is made, at which point they are executed and produced. This means that errors made via the IDE _might_ have different union ordering in their types than errors made on the command line (if you have prior LS results executed that caused a partial load of other types in a different order). But that's likely a fine tradeoff to make; especially since the order already differed depending on how files were included/checked in the IDE, anyway.

Do note: Only diagnostics which depended upon the old `produceDiagnostics` checker flag become lazy in this PR - further work should probably be applied to make more diagnostics lazy, as there are undoubtedly heaps of other diagnostic code that was unconditionally done, which could probably be made lazy to improve LS performance slightly.

Tl;DR: With this change, there can be only one checker. Use `addLazyDiagnostic` to register diagnostic-specific work in the checker.

Fixes #27997

### CODE REVIEW

**Comment by weswigham** [comment]
I don't _expect_ any command-line type changes, so the following should all duplicate `master`'s current state:

@typescript-bot perf test this
@typescript-bot test this
@typescript-bot run dt
@typescript-bot user test this

**Comment by weswigham** [comment]
@typescript-bot pack this

File: tests/baselines/reference/findAllRefs_importType_js.1.baseline.jsonc [hunk 1]
```diff
@@ -15,7 +15,7 @@
       "containerName": "",
       "fileName": "/a.js",
       "kind": "local class",
-      "name": "(local class) C",
+      "name": "(local class) C\nmodule C",
```
Comment (weswigham) [inline tests/baselines/reference/findAllRefs_importType_js.1.baseline.jsonc:18]: This new result is more correct (it accurately reflects the merge done on the symbol), however the "new" result isn't actually stable - symbol merges are ad-hoc in the checker, and the merge that produces this result is only created when the symbol for the _source file_ is pulled on (as by diagnostic production), but not when the local _class_ is pulled on (you can probably trigger the same inconsistency today by requesting all refs for the file followed by all refs for the class)! This is probably an amazingly subtle bug somewhere in JS checking that's been revealed by this diff, and fixing it is a separate issue. (The fix should involve making the merge no matter which entrypoint requests the symbol.)

File: src/services/codefixes/convertFunctionToEs6Class.ts [hunk 2]
```diff
@@ -71,18 +56,31 @@ namespace ts.codefix {
...
+            }
+
+            // all instance members are stored in the "member" array of symbol (done last so instance members pulled from prototype assignments have priority)
+            if (symbol.members) {
+                symbol.members.forEach((member, key) => {
+                    if (key === "constructor" && member.valueDeclaration) {
+                        const prototypeAssignment = symbol.exports?.get("prototype" as __String)?.declarations?.[0]?.parent;
+                        if (prototypeAssignment && isBinaryExpression(prototypeAssignment) && isObjectLiteralExpression(prototypeAssignment.right) && some(prototypeAssignment.right.properties, isConstructorAssignment)) {
+                            // fn.prototype = { constructor: fn }
+                            // Already deleted in `createClassElement` in first pass
```
Comment (weswigham) [inline src/services/codefixes/convertFunctionToEs6Class.ts:75]: Likewise, convertToES6Class had an issue where a js prototype property merge sometimes happens and sometimes does not (and it breaks when the merge _does_ occur in `main`). Since I didn't like crashing fourslash tests, in lieu of fixing the underlying issue (making the merge always occur regardless of what symbol is requested as an entrypoint), I've hardened the implementation to work when a symbol appears as both a `.prototype` member _and_ has been merged into the actual `symbol.members` set (and, coincedentally, moved static members to before instance members in the generated class, which I much prefer). Just like with the only-sometimes-merged js baseline symbol, I think fixing that (existing) bug should be a separate issue.

File: tests/baselines/reference/findAllRefs_importType_js.1.baseline.jsonc (hunk 1 above)
Comment (weswigham) [reply tests/baselines/reference/findAllRefs_importType_js.1.baseline.jsonc:18]: Opened https://github.com/microsoft/TypeScript/issues/48197 to track the issue.

**Comment by gabritto** [comment]
@typescript-bot pack this

**Comment by gabritto** [comment]
The fuzzer ran yesterday (March 22) with this PR's code instead of the usual TypeScript package, and I don't see any new issues/crashes reported.

**Comment by weswigham** [comment]
Oh boy! @RyanCavanaugh does that mean I can sync this (again) and merge it? Reviews appreciated~

**Comment by weswigham** [comment]
Oh, also, @gabritto, does the fuzzer record memory used for each project it fuzzes in its logs somewhere? It'd be cool to know the before/after memory usage stats.

**Review by sheetalkamat — APPROVED** [review]

**Comment by weswigham** [comment]
cc @alexeagle since you were interested in the original issue, many moons ago 👀

**Comment by alexeagle** [comment]
Does that mean we can someday have custom diagnostics configured in `tsc` like http://tsetse.info?

**Comment by weswigham** [comment]
Uhh, it means you should no longer have to worry about the two-checker overhead when doing both diagnostic reporting and other LS operations.
